src/process.rs

```diff
@@ -637,15 +636,333 @@ mod unix {
+                                // this is how the env var protocol for passing sockets is
+                                // defined as FDs are actually just integer values.
+                                for fd in SD_LISTEN_FDS_START..SD_LISTEN_FDS_START + (num_fds as RawFd) {
+                                    match SocketInfo::from_fd(fd) {
+                                        Ok(socket_info) => {
+                                            log::trace!("Received socket file descriptor {} via SystemD LISTEN_FDS env var: type={}, address={}",
+                                                socket_info.raw_fd, socket_info.socket_type, socket_info.socket_addr);
+                                            fds.push(socket_info);
+                                        }
+                                        Err(err) => log::warn!("SystemD LISTEN_FDS provided file descriptor is not usable: {err}"),
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+
+            if unset_environment {
+                log::trace!("Unsetting SystemD LISTEN_PID and LISTEN_FDS environment variables.");
+                std::env::remove_var("LISTEN_PID");
+                std::env::remove_var("LISTEN_FDS");
+            }
+
+            Self { fds }
+        }
+
+        /// Were socket descriptors passed to us via the environment?
+        ///
+        /// Returns false if not, true otherwise.
+        pub fn is_empty(&self) -> bool {
+            self.fds.is_empty()
+        }
+
+        /// Did the environment contain a UDP socket descriptor for
+        /// the specified address?
+        ///
+        /// Returns true if so, false otherwise.
+        pub fn has_udp(&self, addr: &SocketAddr) -> bool {
+            self.fds
+                .iter()
+                .any(|v| v.socket_type == SocketType::Udp && v.socket_addr == *addr)
+        }
+
+        /// Did the environment contain a TCP socket descriptor for
+        /// the specified address?
+        ///
+        /// Returns true if so, false otherwise.
+        pub fn has_tcp(&self, addr: &SocketAddr) -> bool {
+            self.fds
+                .iter()
+                .any(|v| v.socket_type == SocketType::Tcp && v.socket_addr == *addr)
+        }
+
+        /// Returns the specified UDP socket, assuming it was supplied
+        /// to us via the environment.
+        ///
+        /// If found, removes the file descriptor from the collection, sets
+        /// the FD_CLOEXEC flag on the file descriptor and returns it as the
+        /// Rust type Some(UdpSocket).
+        ///
+        /// Subsequent attempts to remove the same UDP socket, or any other
+        /// non-existing socket, will return None.
+        pub fn take_udp(&mut self, addr: &SocketAddr) -> Option<UdpSocket> {
```

Comment:
The `addr` argument needs some more description, specifically, it is not entirely obvious that you expect the “sockname,” i.e., the address the socket is bound to locally. It makes sense to expect that, but I had to go check the source to make sure.